lib/spack/spack/variant.py

```diff
@@ -35,36 +37,77 @@
 special_variant_values = [None, "none", "*"]
 
 
+class ValidValue:
+    def __init__(self, values):
+        self._values = values
+
+    def __call__(self, v):
+        return v in self._values
+
+    def __eq__(self, other):
+        return self._values == other._values
```

Comment:
`single_value_validator` has been part of equality comparison for `Variant` since 67cd92e6a32, so this is just ensuring that we can compare for equality in a sensible way.  To implement the PR here I did a lot of analysis (e.g. traversing and comparing all variant definitions) and this was helpful for doing that.  I think this deserves a revisit but not sure if we should change it now.

Reply:
I actually removed `__ne__`, `__eq__`, and the whole `ValueValidator` class. Much cleaner now.